### PR TITLE
Migrate to typescript

### PR DESCRIPTION
**Pull request by stefantds**
This PR is migrating the library to typescript **almost** without any other changes (the changes introduced are highlighted by comments).
As a follow-up we should continue making the library more typescript-y (e.g. replace multiple functions `decrypt[Object|String|etc]` with union types. This was not in scope for this refactoring.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,35 +1,19 @@
...
+    parser: '@typescript-eslint/parser',
     parserOptions: {
         ecmaVersion: 8,
         sourceType: 'module',
     },
     plugins: [
         'import',
+        '@typescript-eslint'
     ],
-    rules: {
```
Comment (stefantds) [inline .eslintrc.js:11]: I used this occasion to simplify the linting rules. If there are opipnions about custom rules I'm happy to add them back but the recommended values are quite a good start.

File: .eslintrc.js [hunk 2]
```diff
@@ -1,35 +1,19 @@
 module.exports = {
-    extends: ['eslint:recommended', 'airbnb-base'],
```
Comment (stefantds) [inline .eslintrc.js:2]: We could use the airbnb typescript rules set but I am slightly in favour of the more official rules.

File: karma.conf.js [hunk 3]
```diff
@@ -6,19 +6,13 @@ module.exports = function (config) {
...
-            'test/**/*.js',
+            'source/**/*.ts',
+            'test/**/*.ts',
         ],
 
         preprocessors: {
-            'source/**/*.js': ['browserify'],
-            'test/**/*.js': ['browserify'],
-        },
-        browserify: {
```
Comment (stefantds) [inline karma.conf.js:17]: I'm not sure if there is anything still requiring this step.

File: source/index.ts [hunk 4]
```diff
@@ -59,7 +60,7 @@ export {
     exportPublicKeyToSPKI,
     importSymKeyFromBase64,
     exportSymKeyToBase64,
-    KEYTYPES as keyTypes,
+    D4LKeyTypes as D4LKeyTypes,
```
Comment (stefantds) [inline source/index.ts:63]: this is a small breaking change (which we cold easily revert if needed)

File: source/key.ts [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+const DECRYPT = 'decrypt';
+
+enum D4LKeyTypes {
+    COMMON_KEY = 'ck',
+    DATA_KEY = 'dk',
+    MAIN_KEY = 'mk',
+    PASSWORD_KEY = 'pk',
+    ATTACHMENT_KEY = 'ak',
+    TAG_ENCRYPTION_KEY = 'tek',
+    APP_PRIVATE_KEY = 'apriv',
```
Comment (stefantds) [inline source/key.ts:24]: here is another breaking change as I wanted to transform these values to an enum. it used to be:
```
    APP: {
        PRIVATE_KEY: 'apriv',
        PUBLIC_KEY: 'apub',
    },
    USER: {
        PRIVATE_KEY: 'upriv',
        PUBLIC_KEY: 'upub',
    },
```

File: source/key.ts [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+        )
+        .then(key => exportKey(key, D4LKeyTypes.PASSWORD_KEY));
+};
+
+/**
+ * Creates a random public-private key pair
+ * @param {AsymKeyTypes} type - type of the key pair (app/user)
+ * @returns {Promise<D4LAsymKeysPair>} Resolves to an object containing a public and a private key as d4lKey objects
+ */
+const generateAsymKeyPair = (type: AsymKeyTypes): Promise<D4LAsymKeysPair> =>
```
Comment (stefantds) [inline source/key.ts:326]: I changed the type of the parameter here as it was depending on the enum-objects that I removed (it was also wrongly documented):
```
@param {String} type - type of the key pair 'A' for App, 'U' for User
[...]
            exportKey(keyPair.publicKey, type.PUBLIC_KEY),
            exportKey(keyPair.privateKey, type.PRIVATE_KEY),
```

File: source/key.ts [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+        { name: 'PBKDF2' },
+        false,
+        ['deriveKey', 'deriveBits'],
+    )
+        .then(key =>
+            crypto.subtle.deriveKey(
+                {
+                    name: 'PBKDF2',
+                    salt,
+                    iterations: ITERATIONS,
```
Comment (ibihim) [inline source/key.ts:309]: Nit: Makes me wonder if it makes sense to have `ITERATIONS` is too generic as a name. Maybe defining a whole PBKDF2 object would be cool... or not? :smile:  Like in a JS file where you have a quick overview of all the ciphers we use... and maybe an indirection on top so we can easily... sorry... drifting into symbolic links :laughing:

File: source/key.ts [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+
+/**
+ * Export public CryptoKey to base64 encoded SPKI.
+ *
+ * @param {CryptoKey} cryptoKey - the key that should be exported
+ * @returns {Promise<b64>} Resolves to the key as a base64 encoded String.
+ */
+const exportPublicKeyToSPKI = (cryptoKey: CryptoKey): Promise<b64> =>
+    crypto.subtle.exportKey('spki', cryptoKey)
+        .then(SPKI => new Uint8Array(SPKI))
```
Comment (ibihim) [inline source/key.ts:76]: Nit: `SPKI` is not a constant so lower case?

File: source/key.ts (hunk 5 above)
Comment (ibihim) [reply source/key.ts:24]: Your idea adds structure :+1:

**Review by ibihim — COMMENTED** [review]
Nice effort!

File: .eslintrc.js (hunk 2 above)
Comment (radibit) [reply .eslintrc.js:2]: Yes, completely agree 👍

File: karma.conf.js (hunk 3 above)
Comment (radibit) [reply karma.conf.js:17]: It should be needed especially in this Karma related configuration.

File: package.json [hunk 9]
```diff
@@ -1,56 +1,59 @@
...
+        "@babel/core": "^7.5.5",
+        "@babel/polyfill": "^7.0.0-beta.54",
+        "@babel/preset-env": "^7.5.5",
+        "@types/base64-js": "^1.3.0",
+        "@types/mocha": "^8.2.3",
+        "@types/pbkdf2": "^3.0.0",
+        "@typescript-eslint/eslint-plugin": "^4.28.3",
+        "@typescript-eslint/parser": "^4.28.3",
+        "babel-eslint": "^8.2.6",
+        "babelify": "^9.0.0",
```
Comment (radibit) [inline package.json:42]: Do we still need `babelify` considering the changes at https://github.com/d4l-data4life/js-crypto/pull/18/files#diff-25bdb46a5428d318648ec9eb46a99c0268d8c96b672c46e2247a685bc3e384fbL20?

File: source/key.ts [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+    crypto.subtle.generateKey(
+        RSA_OAEP,
+        true,
+        [ENCRYPT, DECRYPT],
+    )
+        .then(keyPair => Promise.all([
+            exportKey(
+                keyPair.publicKey, type === AsymKeyTypes.APP ?
+                D4LKeyTypes.APP_PUBLIC_KEY :
+                D4LKeyTypes.USER_PUBLIC_KEY) as Promise<D4LKeyPublic>,
```
[the remark this answers is not in the document]
Comment (radibit) [reply source/key.ts:342]: This misalignment will be resolved as soon as we introduce Prettier to this project.

**Review by radibit — COMMENTED** [review]
Great work on this PR especially considering the goal to keep it as lean as possible for a review and step-by-step code refactoring in series of PR 👏 

I've left several minor comments aside from what has been already addressed from the previous reviewer.

Would be nice to keep track of the breaking changes outside of the PRs so we can easily add them later to the final release. Maybe a wiki page or section in the readme?

As a follow up PR, I'd add Prettier to the project so we can unify its code formatting with the rest of the projects.

File: karma.conf.js (hunk 3 above)
Comment (stefantds) [reply karma.conf.js:17]: @radibit ok, thanks!

File: source/index.ts [hunk 11]
```diff
@@ -39,12 +40,12 @@ import {
 } from './symmetric';
 import * as algorithms from './algorithms';
 
-const crypto = window.crypto || window.msCrypto;
+const crypto = window.crypto || (window['msCrypto'] as Crypto);
 
-const hash = (message, alg = 'SHA-512') => crypto.subtle.digest(alg, Buffer.from(message))
+const hash = (message: ArrayBuffer, alg = 'SHA-512'): Promise<b64> => crypto.subtle.digest(alg, Buffer.from(message))
```
[the remark this answers is not in the document]
Comment (stefantds) [reply source/index.ts:45]: I agree, but it's not something that I introduced and not in scope for this refactoring. Will keep in mind for the next steps.

File: source/utils.ts [hunk 12]
```diff
@@ -7,37 +8,39 @@ const textDecoder = new TextDecoder();
...
 
-const convertBlobToArrayBufferView = blob =>
+const convertBlobToArrayBufferView = (blob: Blob): Promise<Uint8Array> =>
     new Promise((resolve) => {
         const fileReader = new FileReader();
         fileReader.onload = (event) => {
-            resolve(new Uint8Array(event.target.result));
+
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            resolve(new Uint8Array(event.target.result as any)); // TODO check the type mismatch here, remove "any"
```
[the remark this answers is not in the document]
Comment (stefantds) [reply source/utils.ts:38]: I agree, but it's not something that I introduced and not in scope for this refactoring. Will keep in mind for the next steps.

File: source/key.ts (hunk 7 above)
Comment (stefantds) [reply source/key.ts:309]: yeah, makes sense. Will reconsider the whole const suggestions altogether in a future PR.

File: source/key.ts [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+const generateAsymKeyPair = (type: AsymKeyTypes): Promise<D4LAsymKeysPair> =>
+    // Parameters:
+    // 1. Asymmetric Encryption algorithm name and its requirements
+    // 2. Boolean indicating extractable which indicates whether or not the raw keying
+    //    material may be exported by the application
+    //    (http://www.w3.org/TR/WebCryptoAPI/#dfn-CryptoKey-slot-extractable)
+    // 3. Usage of the keys. (http://www.w3.org/TR/WebCryptoAPI/#cryptokey-interface-types)
+    crypto.subtle.generateKey(
+        RSA_OAEP,
+        true,
```
[the remark this answers is not in the document]
Comment (stefantds) [reply source/key.ts:335]: yeah, although on this we have more issues with how we handle the extractable property of the WebCrypto keys/ E.g. it is confusing and hidden that for some keys we use extractable while for others unextractable keys. This should be left to the client to decide, this library should have no idea how these keys are used.

[the remark this answers is not in the document]
Comment (stefantds) [reply source/key.ts:335]: (nevertheless, the suggestion improves a lot the readability so I will add it already)

**Comment by stefantds** [comment]
@radibit thanks for the review, I addressed the comments (except the constants, as it goes beyond the scope of this PR in my opinion).
Please take a look again.
I still have to see why the build is failing, maybe you already know what is missing there?

**Comment by radibit** [comment]
@stefantds Regarding the constants 👍 

About the failing build, it looks like the `typescript-eslint` might need to explicitly exclude of checking any files under `node_modules`. You can try adding a `.eslintignore` rule for that directory. Reference GH thread https://github.com/facebook/create-react-app/issues/8936#issuecomment-674060591.

**Comment by stefantds** [comment]
@radibit the issue was actually what it was stated: typescript was not installed. I fixed the issue by adding typescript as a dev dependency. I missed that as I had typescript installed globally.

**Review by radibit — APPROVED** [review]
Looks good 🥇

On the TypeScript dependency, indeed it's a required one for any of the plugins that relay on it like in the case with linting a `*.ts` files 😅

**Pull request merged**